libarchive/archive_write_set_format_xar.c

```diff
@@ -3252,4 +3290,120 @@ getalgname(enum sumalg sumalg)
 	}
 }
 
+#if HAVE_LIBXML_XMLWRITER_H
+
+#define BAD_CAST_CONST (const xmlChar *)
+
+struct xml_writer {
+	xmlTextWriterPtr writer;
+	xmlBufferPtr bp;
+	unsigned int indent;
+};
+
+static int
+xml_writer_create(struct xml_writer **pctx) {
+	struct xml_writer *ctx = calloc(1, sizeof(struct xml_writer));
+	if (ctx == NULL) {
+		return (-1);
+	}
+
+	ctx->bp = xmlBufferCreate();
+	if (ctx->bp == NULL) {
+		free(ctx);
+		return (-1);
+	}
+
+	ctx->writer = xmlNewTextWriterMemory(ctx->bp, 0);
+	if (ctx->writer == NULL) {
+		xmlBufferFree(ctx->bp);
+		free(ctx);
+		return (-1);
```

Comment:
The caller will print a bunch of `-1` messages now on error, I'd suggest swapping these with ENOMEM.

Reply:
Great call. Thanks!

Reply:
So, thinking about this one more: the original code did not indicate failure with `ENOMEM` here either. Just that "it failed"!

I'm on the fence about switching these ones over, since to uphold the consistent contract that negative return values are errors it would need to be `-ENOMEM` and then we're in ever-hotter water because that's really not a useful error code _either_.
The internal consumer would dutifully report it as `-12` (or whatever)... but at the end of the day it would still be a fatal failure.

Reply:
The original code returns an `ARCHIVE_ERRNO_MISC` with "xmlNewTextWriterMemory() couldn't create xml writer". The new  `ARCHIVE_ERRNO_MISC` with  "xml_writer_create() failed: -1".

Both aren't particularly useful, although since this PR is adding an abstraction it shouldn't really change it IMHO.